src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.cpp

```diff
@@ -188,16 +187,56 @@ void SequenceViewAnnotatedRenderer::drawAnnotation(QPainter& p, const QSize& can
-                }
             }
             drawAnnotationConnections(p, a, as, displaySettings, canvasSize, visibleRange);
-        } else if (isRestrictionSite && simple && displaySettings.displayCutSites) {
-            p.setPen(borderPen);
-            drawCutSite(p, aData, r, QRect(), as->color, canvasSize, visibleRange);
+        }
+    }
+    // draw restriction sites
+    /**
+     * Calculates restriction cuts positions depends on visible area and calls drawCutSite.
+     *
+     * \param reg Region of the annotation, which sites should be drawn.
+     * \param drawDirect Draw direct cut if true and do not draw if false. Default value is true.
+     * \param drawComplement Draw complement cut if true do not draw if false. Default value is true.
+     */
+    auto drawCutSiteLambda = [this, &p, &visibleRange, &a, &as, &availableHeight, &canvasSize, selected, &aData](const U2Region& reg, bool drawDirect = true, bool drawComplement = true) {
+        if (reg.intersects(visibleRange)) {
+
+            const U2Region visibleLocation = reg.intersect(visibleRange);
+            const U2Region y = getAnnotationYRange(a, 0, as, availableHeight);
+            CHECK(y.startPos >= 0, );
+
+            const int x1 = posToXCoord(visibleLocation.startPos, canvasSize, visibleRange);
+            const int x2 = posToXCoord(visibleLocation.endPos(), canvasSize, visibleRange);
+
+            const int rw = qMax(selected ? MIN_SELECTED_ANNOTATION_WIDTH : MIN_ANNOTATION_WIDTH, x2 - x1);
+            SAFE_POINT(rw > 0, "Negative length of annotationYRange", );
+
+            const QRect annotationRect(x1, y.startPos, rw, y.length);
+            if (annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES) {
+                drawCutSite(p, aData, reg, annotationRect, as->color, canvasSize, visibleRange, drawDirect, drawComplement);
+            }
+        } else {
+            drawCutSite(p, aData, reg, QRect(), as->color, canvasSize, visibleRange);
+        }
+    };
+
+    if (isRestrictionSite && displaySettings.displayCutSites) {
+        p.setPen(borderPen);
+        const bool isCircular = ctx->getSequenceObject()->isCircular();
+        auto seqLength = ctx->getSequenceLength();
+        if (isCircular && U1AnnotationUtils::isAnnotationContainsJunctionPoint(a, seqLength)) {
+            auto merged = U1AnnotationUtils::mergeAnnotatedRegionsAroundJunctionPoint(location, seqLength);
+            SAFE_POINT(merged.size() == 1, "Unexpected merged enzyme region", );
+
+            auto regionPair = merged.front();
+            // draw cuts separately,
+            // because they are located one on the 3' end of the direct sequence,
+            // and the other one on the 3' end of reverse-complementary sequence
+            drawCutSiteLambda(regionPair.first, true, false);
+            drawCutSiteLambda(regionPair.second, false, true);
+        } else if (location.size() == 1) {
```

Comment:
Here I found a problem (thanks to GUI tests) I need to discuss - it could be some possible bug or something. Previously, this line was just `else` and line above contained `SAFE_POINT`:  
`SAFE_POINT(location.size() == 1, ...);`  
But, I faced a problem here, **scenario**:

1.  Open "data\samples\Genbank\PBR322.gb" (it is circular).
2. Open the "Restriction sites" dialog and choose the only enzyme "EcoRI".
Expected: enzyme on the junction point has been found.
3. Uncheck "Mark as circular".
The above mentioned `SAFE_POINT` has been triggered.

Logically, it shouldn't happen - if sequence was not circular, than restriction enzyme just wouldn't be found. The problem is here - when we uncheck "Mark as circular", the `Auto-annotations update` task being triggered to update all annotations. This task doesn't work instantly and when this task is still in process, old restriction enzymes are trying to be shown, and the site, which was located on junction point just being split to two parts and trying to be shown as "joined". As far as we anticipate, that restriction enzymes annotations are **always** inseparable in case of non-circular view, `SAFE_POINT` is being triggered. 

You thing it could be a bug and we need to prevent somehow splitting restriction enzymes to two joined parts or not?

Reply:
Here is what I think

1) We should show enzymes correctly for circular sequences.
2) We have no other way to draw annotations on circular sequences that cross junction point but only to split them into 2
3) The above makes enzymes split into 2 - we have no option

The code that deals with enzymes should be ready for this situation: allow 2 regions of kind (P1...LEN) & (0...P2) and handle them correctly.

This is what you already do in this patch and this is good.